# HG changeset patch
# User Rik <rik@octave.org>
# Date 1371668166 25200
#      Wed Jun 19 11:56:06 2013 -0700
# Node ID 5ec3f4aea91cfe1f854cd65ccaf73ccff1b64656
# Parent  8d188159ce5f1728a0f84e5fe7c5f28ffbe56478
fplot.m: Fix bug when FN is a constant function (bug #39287)

* scripts/plot/fplot.m: Check for constant value function after
feval call.  Add input validation tests.

diff --git a/scripts/plot/fplot.m b/scripts/plot/fplot.m
--- a/scripts/plot/fplot.m
+++ b/scripts/plot/fplot.m
@@ -74,16 +74,22 @@ function fplot (fn, limits, n = 0.002, f
   if (n != fix (n))
     tol = n;
     x0 = linspace (limits(1), limits(2), 5)';
     y0 = feval (fn, x0);
     err0 = Inf;
     n = 8;
     x = linspace (limits(1), limits(2), n)';
     y = feval (fn, x);
+   
+    if (! size_equal (x0, y0))
+      ## FN is a constant value function
+      y0 = repmat (y0, size (x0));
+      y = repmat (y, size (x));
+    endif
 
     while (n < 2 .^ 20)
       y00 = interp1 (x0, y0, x, "linear");
       err = 0.5 * max (abs ((y00 - y) ./ (y00 + y))(:));
       if (err == err0 || err < tol)
         break;
       endif
       x0 = x;
@@ -119,8 +125,16 @@ endfunction
 %!demo
 %! clf;
 %! fplot ('cos', [0, 2*pi]);
 
 %!demo
 %! clf;
 %! fplot ('[cos(x), sin(x)]', [0, 2*pi]);
 
+%% Test input validation
+%!error fplot (1)
+%!error fplot (1,2,3,4,5)
+%!error <LIMITS must be a real vector> fplot (@cos, [i, 2*i])
+%!error <LIMITS must be a real vector with 2 or 4> fplot (@cos, [1])
+%!error <LIMITS must be a real vector with 2 or 4> fplot (@cos, [1 2 3])
+%!error <FN must be a function handle> fplot (1, [0 1])
+
